client/go/outline/reporting/reporting_client.go

```diff
@@ -0,0 +1,271 @@
+						Value: cookieParts[1],
+					})
+					fmt.Println("Cookie found:", cookie)
+				}
+			}
+		}
+	}
+	// Check if the URL is already in the list
+	exists := false
+	for _, existingURL := range urls {
+		if existingURL.String() == u.String() {
+			exists = true
+			break
+		}
+	}
+	if !exists {
+		urls = append(urls, u)
+	}
+	jar.SetCookies(u, newCookies)
+	// Save new cookies to file
+	if err := saveCookies(urls, jar, cookiesFile); err != nil {
+		return fmt.Errorf("failed to save cookies: %w", err)
+	}
+
+	return nil
+}
+
+// CookieData represents the structure for saving cookies, including the URL
+type CookieData struct {
+	URL     string         `json:"url"`
+	Cookies []CookieDetail `json:"cookies"`
+}
+
+type CookieDetail struct {
+	Name   string `json:"name"`
+	Value  string `json:"value"`
+	Domain string `json:"domain"`
+	Path   string `json:"path"`
+}
+
+func saveCookies(urls []*url.URL, jar http.CookieJar, filename string) error {
+	// Retrieve all cookies from the jar
+	cookieDataMap := make(map[string][]CookieDetail)
+
+	for _, urlObj := range urls {
+		cookies := jar.Cookies(urlObj)
+		var cookieDetails []CookieDetail
+		for _, cookie := range cookies {
+			cookieDetails = append(cookieDetails, CookieDetail{
+				Name:   cookie.Name,
+				Value:  cookie.Value,
+				Domain: cookie.Domain,
+				Path:   cookie.Path,
+			})
+		}
+
+		cookieDataMap[urlObj.String()] = cookieDetails
+	}
+
+	// Save the cookies data to a file
+	file, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	encoder := json.NewEncoder(file)
+	if err := encoder.Encode(cookieDataMap); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func loadCookies(jar http.CookieJar, filename string) ([]*url.URL, error) {
+	// Read cookies from the file
+	file, err := os.Open(filename)
+	if os.IsNotExist(err) {
+		return []*url.URL{}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var cookieDataMap map[string][]CookieDetail
+	decoder := json.NewDecoder(file)
+	if err := decoder.Decode(&cookieDataMap); err != nil {
+		return nil, err
+	}
+
+	var urls []*url.URL
+	// Convert the loaded data back into http.Cookie objects and load them into the jar
+	for urlStr, cookieDetails := range cookieDataMap {
+		parsedURL, err := url.Parse(urlStr)
+		if err != nil {
+			return nil, err
+		}
+
+		var cookies []*http.Cookie
+		for _, c := range cookieDetails {
+			cookies = append(cookies, &http.Cookie{
+				Name:   c.Name,
+				Value:  c.Value,
+				Domain: c.Domain,
+				Path:   c.Path,
+			})
+		}
+
+		// Set the cookies in the jar for the specified URL
+		jar.SetCookies(parsedURL, cookies)
+		urls = append(urls, parsedURL)
+	}
+
+	return urls, nil
+}
+
+// StartReporting calls the Report function every 10 seconds
+func StartReporting(tcp transport.StreamDialer, ur *config.UsageReporter) {
+	if !ur.EnableCookies {
+		return
+	}
+	ticker := time.NewTicker(ur.Interval)
+	defer ticker.Stop()
+
+	for {
```

Comment:
Is this dead loop? Shall we have some logic for shutting down the reporting after user disconnects?

Reply:
If we have to, we might need to avoid create multiple dead loops when user connects to different servers?

Reply:
Done.